--- hba.conf.orig	2009-02-24 11:37:16.000000000 -0800
+++ hba.conf	2009-02-24 11:37:42.000000000 -0800
@@ -16,12 +16,7 @@
 #   the load will fail with the result that the 32 bit vendor
 #   library is ignored.
 
-org.snia.sample32	/usr/lib/libsample.so	          
-com.qlogic.qla32	/usr/lib/libhbaapiqla.so
-com.emulex.lightpulse32 /usr/lib/libhbaapiemu.so
-com.emulex.lightpulse64	/usr/lib/sparcv9/libhbaapiemu.so
-JniHbaLib		/opt/JNIsnia/Solaris/Jni/32bit/JniHbaLib.so
-JniHbaLib		/opt/JNIsnia/Solaris/Jni/64bit/JniHbaLib.so
+# org.snia.sample32	/usr/lib/libsample.so	          
 
 
 # Set a debug level (depends on compilation switches)
